### PR TITLE
(WIP) Most basic dynamic exposure compensation

### LINKED ISSUE #120: Tracking issue: HDR and friends
We want:

- [X] an HDR framebuffer to draw into 
  - [ ] maybe reduce its size by using `F16` channel format
- [X] filmic tonemapping
- [X] Bloom
- [x] dynamic exposure compensation
- [x] eye adaption (slow adaption of exposure value) 
- [x] Bloom using the dynamically calculated exposure value
- [x] Tonemapping using the dynamically calculated exposure value

### PR DESCRIPTION
**Pull request by gustav1101**
Calculation of average brightness is implemented.
The next step is to reasonably change the exposure level.

Fixes #120

### CODE REVIEW

File: client/src/renderer.rs [hunk 1]
```diff
@@ -363,6 +384,60 @@ impl Renderer {
...
+            .main_level()
+            .first_layer()
+            .into_image(None)
+            .unwrap()
+            .raw_read(&glium::Rect {
+                left: 0,
+                bottom: 0,
+                width: 1,
+                height: 1,
+            });
```
Comment (LukasKalbertodt) [inline client/src/renderer.rs]: dafuq

File: client/src/renderer.rs [hunk 2]
```diff
@@ -372,3 +447,25 @@ struct Vertex {
...
+
+
+
+// ===================================================================
+//                  Brightness Adaption Data Structures
+// ===================================================================
+
+
+fn initialize_luminosity(facade: &GlutinFacade) -> Vec<Texture2d> {
+    let mut lum: Vec<Texture2d> = Vec::with_capacity(10);
```
Comment (LukasKalbertodt) [inline client/src/renderer.rs]: This type annotation shouldn't be necessary, right?

File: client/src/renderer.rs [hunk 3]
```diff
@@ -372,3 +447,25 @@ struct Vertex {
...
+    for i in 0..10 {
+        lum.push(Texture2d::empty_with_format(facade,
+                                              UncompressedFloatFormat::F32F32F32F32,
+                                              MipmapsOption::NoMipmap,
+                                              (2 as u32).pow((9 - i) as u32),
+                                              (2 as u32).pow((9 - i) as u32))
```
Comment (LukasKalbertodt) [inline client/src/renderer.rs]: Whats up with those casts? I think `2.pow(9-i)` should work...

File: client/src/renderer.rs (hunk 1 above)
Comment (jonas-schievink) [inline client/src/renderer.rs]: Yeah this was the thing we needed 4 people to figure out

File: client/src/renderer.rs (hunk 3 above)
Comment (FischerTo) [inline client/src/renderer.rs]: Removing casts results in error: "no method name 'pow' found for type'_' in the current scope"

File: client/src/renderer.rs [hunk 4]
```diff
@@ -59,16 +66,60 @@ impl Renderer {
...
+        let lum_texs = initialize_luminosity(context.get_facade());
+
+        let last_lum = 2.0;
+
+        // let lum_relative_tex = Texture2d::empty_with_format(context.get_facade(),
+        // UncompressedFloatFormat::F32F32,
+        // MipmapsOption::NoMipmap,
+        // resolution.0,
+        // resolution.1)
+        // .unwrap();
```
Comment (LukasKalbertodt) [inline client/src/renderer.rs]: Can this be removed?

File: client/src/renderer.rs [hunk 5]
```diff
@@ -59,16 +66,60 @@ impl Renderer {
...
+
+
         let tonemapping_program = context.load_program("tonemapping").unwrap();
         let bloom_filter_program = context.load_program("bloom_filter").unwrap();
         let bloom_blur_program = context.load_program("bloom_blur").unwrap();
         let bloom_blend_program = context.load_program("bloom_blending").unwrap();
         let shadow_debug_program = context.load_program("shadow_debug").unwrap();
+        let adaption_shrink_program = context.load_program("adaption_shrink").unwrap();
+        // let relative_luminance_program =
+        // context.load_program("relative_luminance").unwrap();
```
Comment (LukasKalbertodt) [inline client/src/renderer.rs]: And what about this?

File: client/src/renderer.rs [hunk 6]
```diff
@@ -226,22 +295,24 @@ impl Renderer {
                                                                          .to_color_attachment()));
         bloom_blur_vert_buffer.clear_color(0.0, 0.0, 0.0, 1.0);
 
+
+
```
Comment (LukasKalbertodt) [inline client/src/renderer.rs]: Random empty lines

File: client/src/renderer.rs [hunk 7]
```diff
@@ -226,22 +295,24 @@ impl Renderer {
...
         let mut uniforms_horz_blur = uniform! {
             //for the first iteration: Use the bloom quad texture, from second iteration on this
             // will change to bloom_vert_tex
-            image: &self.bloom_filter_tex,
+            image: glium::uniforms::Sampler::new(&self.bloom_filter_tex).wrap_function(SamplerWrapFunction::Clamp),
             horizontal: true,
         };
         let uniforms_vert_blur = uniform! {
-            image: &self.bloom_horz_tex,
+            image: glium::uniforms::Sampler::new(&self.bloom_horz_tex).wrap_function(SamplerWrapFunction::Clamp),
```
Comment (LukasKalbertodt) [inline client/src/renderer.rs]: You can write `self.bloom_horz_text.sampled().wrap_function(...)`, it's shorter

File: client/src/renderer.rs [hunk 8]
```diff
@@ -257,7 +328,7 @@ impl Renderer {
             }
             if first_iteration {
                 uniforms_horz_blur = uniform! {
-                    image: &self.bloom_vert_tex,
+                    image: glium::uniforms::Sampler::new(&self.bloom_vert_tex).wrap_function(SamplerWrapFunction::Clamp),
```
Comment (LukasKalbertodt) [inline client/src/renderer.rs]: Same as above

File: client/src/renderer.rs [hunk 9]
```diff
@@ -395,6 +466,86 @@ impl Renderer {
...
+        //
+        //
+        //
+        // try!(rel_luminance_buffer.draw(&self.quad_vertex_buffer,
+        // &self.quad_index_buffer,
+        // &self.relative_luminance_program,
+        // &uniforms,
+        // &Default::default()));
+        //
+        //
```
Comment (LukasKalbertodt) [inline client/src/renderer.rs]: Can this code be removed?

File: client/src/renderer.rs [hunk 10]
```diff
@@ -395,6 +466,86 @@ impl Renderer {
...
+                width: 1,
+                height: 1,
+            });
+
+        let pixel = buf[0][0];
+
+        let avg_luminance = Vector3f::new(pixel.0, pixel.1, pixel.2)
+            .dot(Vector3f::new(0.2126, 0.7152, 0.0722));
+
+        let adapted_luminance = (1.0 / ((avg_luminance + 1.0).log2())).min(6.0).max(1.1);
```
Comment (LukasKalbertodt) [inline client/src/renderer.rs]: This looks crazy... please explain what's going on in the comments.

File: client/src/renderer.rs [hunk 11]
```diff
@@ -395,6 +466,86 @@ impl Renderer {
...
+
+
+
+
```
Comment (LukasKalbertodt) [inline client/src/renderer.rs]: Random empty lines

File: client/src/renderer.rs [hunk 12]
```diff
@@ -404,3 +555,25 @@ struct Vertex {
 }
 
 implement_vertex!(Vertex, in_position, in_texcoord);
+
+
+
+
+
```
Comment (LukasKalbertodt) [inline client/src/renderer.rs]: So many empty lines :fearful:

File: client/src/renderer.rs (hunk 3 above)
Comment (jonas-schievink) [inline client/src/renderer.rs]: Yeah, resolving inherent impl methods requires the self type to be fully known

File: client/src/renderer.rs [hunk 13]
```diff
@@ -404,3 +555,25 @@ struct Vertex {
...
+fn initialize_luminosity(facade: &GlutinFacade) -> Vec<Texture2d> {
+    let mut lum = Vec::with_capacity(10);
+    for i in 0..10 {
+        lum.push(Texture2d::empty_with_format(facade,
+                                              UncompressedFloatFormat::F32F32F32F32,
+                                              MipmapsOption::NoMipmap,
+                                              (2 as u32).pow((9 - i)),
+                                              (2 as u32).pow((9 - i)))
```
Comment (LukasKalbertodt) [inline client/src/renderer.rs]: Do you really need this `as u32`? If so, at least write `2u32` instead :wink:

Comment (FischerTo) [inline client/src/renderer.rs]: We need to cast it to something that let's us use the `pow()` . If we try to use it without casting the compiler complains that there is no such method:
"no method name 'pow' found for type '_' in the current scope"

Comment (jonas-schievink) [inline client/src/renderer.rs]: `2u32` should work

**Pull request merged**